### ISSUE TITLE
Array pattern in destructuring assignment do not typecheck correctly when assigned to an object.

### ISSUE THREAD

**Issue by eyyub** [facebook/flow#733]
``` js
/* @flow */

var [a, b] = {foo:3, b:'lol'}; // no error
(a : string);
(a : number);
(a : void);
(a : any);
(b : void);
(b : any);

var {p} = [1]; // error, ok
var [a, b] : [number | string] = {foo:3, b:'lol'}; // error due to type annotation, ok
```

It seems it comes from [here](https://github.com/facebook/flow/blob/master/src/typing/flow_js.ml#L2873-L2877).
Maybe in [destructuring](https://github.com/facebook/flow/blob/master/src/typing/type_inference_js.ml#L299) we should ensure that the type are "compatible"(good english word ?).

**Comment by kkirby** [comment]
I'm not sure that this is really a bug. Destructing works on objects that are iterators. For example, this is valid JavaScript:

```javascript
let [a,b] = {
  0:'A',
  1:'B',
  [Symbol.iterator]: function*(){
    yield this['0'];
    yield this['1'];
  }
};
console.log(a,b); // Outputs "A B"
```

But in your later example, since an object cannot be an array, an error is thrown.

**Comment by Eyyub** [comment]
I see, at least it should be handled as a special case(actually Flow doesn't support this construct) , but IMHO `var [a, b] = {foo:3, b:'lol'}` should not be allowed.

**Comment by samwgoldman** [comment]
Yeah, we should be able to constrain the RHS to be an iterable in array destructuring assignments.

**Comment by TrySound** [comment]
@samwgoldman This became a more relevant problem since hooks were introduced. I convert tuple api to object and miss some places.

**Comment by kevinbarabash** [comment]
I ran into this today.  The current  behavior results in variables that are typed as `any`.  I should probably code with type coverage always enabled as that would've detected the issue with my code.

**Comment by somewhatabstract** [comment]
See also #2612

**Comment by gkz** [comment]
Fixed in 80ca16f77ba81a3475dd1f7a9f4d9d9e3d71bca8
